### PR TITLE
style: add pre-commit hooks

### PR DESCRIPTION
**Pull request by trim21**
from https://github.com/beancount/beancount/pull/861#issuecomment-2460177363

there are some formatting issue in current codebase, for example, ruff version not locked and ruff didn't run on some file.

This PR add https://github.com/pre-commit/pre-commit to enfore code style, for example, UNIX EOL and ruff linted and formatted code.

also I suggest to add https://github.com/pre-commit/pre-commit-hooks?tab=readme-ov-file#trailing-whitespace, but our `render_source` will generate source with tailing whitespace, and it will failed tests.

### CODE REVIEW

**Comment by dnicolodi** [comment]
This removes white spaces from generated files in the `examples` folder. The content there should probably be updated to the current version of Beancount, but auto formatting generated files does not seem a great idea to me.

**Comment by trim21** [comment]
make sense, I exclude exampels from pre-commit

File: pyproject.toml [hunk 1]
```diff
@@ -50,7 +50,7 @@ line-length = 92
 
 [tool.ruff]
 # Assume Python 3.12.
-target-version = "py312"
+target-version = "py38"
```
Comment (blais) [inline pyproject.toml:53]: Why demote to 3.8?

File: experiments/docs/bean-transaction-types [hunk 2]
```diff
@@ -59,53 +70,54 @@ def main():
...
 
     # Print all renamed account names.
-    print('* Account Names')
+    print("* Account Names")
     print()
-    for account in sorted({account
-                           for accounts in groups.keys()
-                           for account in accounts}):
-        print(account)
+    for a in sorted({a for accounts in groups.keys() for a in accounts}):
```
Comment (blais) [inline experiments/docs/bean-transaction-types]: Did ruff fmt rename the variable? 
I avoid single-character variable names on purpose (Google style).

File: pyproject.toml (hunk 1 above)
Comment (trim21) [reply pyproject.toml:53]: py37 already reach its EOL

File: Makefile [hunk 3]
```diff
@@ -156,6 +156,7 @@ constraints dep-constraints: build/beancount.deps
 # Run the linter on all source code.
 ruff lint:
 	NO_COLOR=1 ruff check .
+	NO_COLOR=1 ruff format .
```
Comment (blais) [inline Makefile:159]: I love it.

File: experiments/docs/bean-transaction-types (hunk 2 above)
Comment (trim21) [reply experiments/docs/bean-transaction-types]: do, ruff didn't. I did this. in this file there is a import account and ruff lint failed here.

**Comment by trim21** [comment]
any chance merge this into rust branch, too?

**Comment by blais** [comment]
Yes, the rust branch is based off of master, so rebasing will add it in. LMDo that now

**Pull request merged**